File: Solutions/23-merge-k-sorted-lists/merge-k-sorted-lists.dart
/**
 * Definition for singly-linked list.
 */

class Solution {
  /// Merge k sorted linked lists into one sorted linked list
  ListNode? mergeKLists(List<ListNode?> lists) {
    if (lists.isEmpty) return null;
    return mergeRange(lists, 0, lists.length - 1);
  }

  /// Helper: merge lists in range [left, right] using divide and conquer
  ListNode? mergeRange(List<ListNode?> lists, int left, int right) {
    if (left == right) return lists[left];
    int mid = left + (right - left) ~/ 2;
    ListNode? l1 = mergeRange(lists, left, mid);
    ListNode? l2 = mergeRange(lists, mid + 1, right);
    return mergeTwoLists(l1, l2);
  }

  /// Merge two sorted linked lists
  ListNode? mergeTwoLists(ListNode? l1, ListNode? l2) {
    ListNode dummy = ListNode(-1);
    ListNode current = dummy;

    while (l1 != null && l2 != null) {
      if (l1.val < l2.val) {
        current.next = l1;
        l1 = l1.next;
      } else {
        current.next = l2;
        l2 = l2.next;
      }
      current = current.next!;
    }

    // Append remaining nodes
    if (l1 != null) current.next = l1;
    if (l2 != null) current.next = l2;

    return dummy.next;
  }
}

File: Solutions/23-merge-k-sorted-lists/merge-k-sorted-lists.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end

# @param {ListNode[]} lists
# @return {ListNode}
def merge_k_lists(lists)
  # Remove nil lists
  heap = lists.compact

  # Edge case: if all lists are empty
  return nil if heap.empty?

  # Dummy node to build result
  dummy = ListNode.new(0)
  current = dummy

  # Sort initial heap based on node values
  heap.sort_by! { |node| node.val }

  while !heap.empty?
    # Take the smallest node
    smallest = heap.shift

    # Attach to result
    current.next = smallest
    current = current.next

    # If there's a next node, add it to heap
    if smallest.next
      heap << smallest.next
      heap.sort_by! { |node| node.val }  # re-sort heap
    end
  end

  return dummy.next
end